src/metabase/lib/util.cljc

```diff
@@ -0,0 +1,100 @@
+(ns metabase.lib.util
+  "TODO -- this should probably be rolled into [[metabase.lib.query]]."
+  (:require
+   [clojure.set :as set]
+   [metabase.lib.options :as lib.options]
+   [metabase.lib.schema :as lib.schema]
+   [metabase.lib.schema.common :as lib.schema.common]
+   [metabase.shared.util.i18n :as i18n]
+   [metabase.util.malli :as mu]))
+
+(defn- native-query->pipeline [query]
+  (merge {:lib/type :mbql/query
+          :type     :pipeline
+          ;; we're using `merge` here instead of threading stuff so the `:lib/` keys are the first part of the map for
+          ;; readability in the REPL.
+          :stages   [(merge (lib.options/ensure-uuid {:lib/type :mbql.stage/native})
+                            (set/rename-keys (:native query) {:query :native}))]}
+         (dissoc query :type :native)))
+
+(defn- inner-query->stages [{:keys [source-query], :as inner-query}]
+  (let [stages     (if source-query
+                     (inner-query->stages source-query)
+                     [])
+        stage-type (if (:native inner-query)
+                     :mbql.stage/native
+                     :mbql.stage/mbql)
+        ;; we're using `merge` here instead of threading stuff so the `:lib/` keys are the first part of the map for
+        ;; readability in the REPL.
+        this-stage (merge (lib.options/ensure-uuid
+                           {:lib/type stage-type})
+                          (dissoc inner-query :source-query))]
+    (conj stages this-stage)))
+
+(defn- mbql-query->pipeline [query]
```

Comment:
I'm confused by this name. It's alluding to the fact that `query` should have `:type` `query`, right?
Having `:lib/type` and `:type` is also somewhat confusing.

Reply:
The QP does a lot of dispatch off of `:type`, so we'll need something there so it knows how to handle a `:type :pipeline` query vs. a `:type :query` or something like that.

`:lib/type` is the key we use for dispatch in MLv2 that tells us what type of map we're dealing with.

We could just use `:type`, but I was worried that it could conflict with other things that might have it -- a separate key was more clearly associated with MLv2, and also safer to strip out if we want to convert things back to QP MBQL (the older version of MBQL currently understood by the QP.

I'll add more documentation to explain things more clearly